components/script/dom/transformunderlyingsource.rs

```diff
@@ -0,0 +1,117 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::rc::Rc;
+
+use dom_struct::dom_struct;
+use js::rust::HandleValue as SafeHandleValue;
+
+use super::bindings::root::DomRoot;
+use super::types::TransformStream;
+use crate::dom::bindings::error::Fallible;
+use crate::dom::bindings::reflector::{DomGlobal, Reflector, reflect_dom_object};
+use crate::dom::bindings::root::Dom;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+#[dom_struct]
+/// <https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaulttee>
+pub(crate) struct TransformUnderlyingSource {
+    reflector_: Reflector,
+    stream: Dom<TransformStream>,
+    #[ignore_malloc_size_of = "Rc is hard"]
+    start_promise: Rc<Promise>,
+}
```

Comment:
I'm not sure if this struct is necessary, you could just put `stream` and `start_promise` in `UnderlyingSinkType::Transform`. The algo's call into the stream so they're small enough to just put in the writable default controller. 

That is what I did in https://github.com/servo/servo/pull/36181